bootloader/src/pyi_archive.c

```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
 #endif /* ifdef _WIN32 */
 }
 
+/*
+ * Try to find the pydata section in an ELF executable.
+ * Returns the end of the section if found.
+ */
+static int
+findElfSection(ARCHIVE_STATUS * const status)
+{
+#if !defined(WIN32) && !defined(__APPLE__)
+    char head[64];
+    uint16_t endianess_bytes = 0x0201;
+    int8_t bits;
+    uint64_t e_shoff;
+    uint16_t e_shentsize;
+    uint16_t e_shnum;
+    uint16_t e_shstrndx;
+    char *section;
+    uint64_t shstr_offset;
+    uint64_t shstr_size;
+    char *shstr;
+    uint32_t section_idx;
+    uint64_t pydata_offset;
+    uint64_t pydata_size;
+
+    fseek(status->fp, 0, SEEK_SET);
+    if (fread(head, 64, 1, status->fp) != 1) {
+        return -1;
+    }
+
+    /* Check if ELF magic present */
+    if (head[0] != 0x7F || head[1] != 'E' || head[2] != 'L' || head[3] != 'F') {
+        VS("LOADER: %s is not an ELF binary\n", status->archivename);
+        return -1;
+    }
+
+    /* Ensure endianess of file matches */
+    if (head[0x05] != *(uint8_t*)&endianess_bytes) {
+        VS("LOADER: %s endianess does not match CPU\n", status->archivename);
```

Comment:
This assumes that the CPU is little-endian. I don't know if PyInstaller works correctly on big-endian platforms at the  moment, but this should either check the endianness of the CPU or simply say that big-endian ELF files are not supported.

Reply:
@lkollar See above. It actually does work on both little and big endian ARM systems.